### PR TITLE
Fix ending span in Ktor plugin

### PR DESCRIPTION
**Pull request by marychatte**
In the documentation written next:

> HTTP client spans SHOULD end sometime after the HTTP response headers are fully read (or when they fail to be read). This may or may not include reading the response body. (https://opentelemetry.io/docs/specs/semconv/http/http-spans/#http-client-span-duration)

Right now, the span ends when headers are fully read. But in this case, users sometimes measure incorrect time of request, and this difference can be extremely big in the case of log-running calls, for example, SSE.

The current PR suggests changing the end of the span after reading the body.

### CODE REVIEW

File: testing-common/src/main/java/io/opentelemetry/instrumentation/testing/junit/http/HttpClientTestServer.java [hunk 1]
```diff
@@ -99,6 +101,24 @@ protected void configure(ServerBuilder sb) throws Exception {
             "/read-timeout",
             (ctx, req) ->
                 HttpResponse.delayed(HttpResponse.of(HttpStatus.OK), Duration.ofSeconds(20)))
+        .service(
+            "/long_body_receiving",
```
Comment (laurit) [inline testing-common/src/main/java/io/opentelemetry/instrumentation/testing/junit/http/HttpClientTestServer.java:105]: other endpoints use `-` as word separator. Perhaps name it `long-request`?

File: instrumentation/ktor/ktor-2.0/testing/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/client/AbstractKtorHttpClientTest.kt [hunk 2]
```diff
@@ -71,4 +77,24 @@ abstract class AbstractKtorHttpClientTest : AbstractHttpClientTest<HttpRequestBu
       }
     }
   }
+
+  @Test
+  fun checkSpanEndsAfterBodyReceived() {
```
Comment (laurit) [inline instrumentation/ktor/ktor-2.0/testing/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/client/AbstractKtorHttpClientTest.kt:82]: Would it make sense to move this to `AbstractHttpClientTest` to ensure that other http client instrumentations also behave the same way?

File: instrumentation/ktor/ktor-2.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/client/KtorClientTracing.kt [hunk 3]
```diff
@@ -83,23 +86,21 @@ class KtorClientTracing internal constructor(
       }
     }
 
+    @OptIn(InternalCoroutinesApi::class)
     private fun installSpanEnd(plugin: KtorClientTracing, scope: HttpClient) {
       val endSpanPhase = PipelinePhase("OpenTelemetryEndSpan")
       scope.receivePipeline.insertPhaseBefore(HttpReceivePipeline.State, endSpanPhase)
```
Comment (laurit) [inline instrumentation/ktor/ktor-2.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/client/KtorClientTracing.kt:92]: would changing the phase to `HttpReceivePipeline.After` also work?

**Comment by marrek13** [comment]
Will this PR also solve the issue with missing WebSocket spans? We currently struggle to see the root span of a WebSocket call. According to the OTel agent's debug logs, nothing was sent related to WS.

**Comment by marychatte** [comment]
@marrek13 Can you please add more details: is it on client or on server? And if possible, create a ticket in Youtrack Ktor https://youtrack.jetbrains.com/newIssue?project=KTOR

**Comment by marrek13** [comment]
> @marrek13 Can you please add more details: is it on client or on server? And if possible, create a ticket in Youtrack Ktor https://youtrack.jetbrains.com/newIssue?project=KTOR

It's on server side. The root span is not sent, I'll create a ticket with details tomorrow and cross post it here.

File: instrumentation/ktor/ktor-2.0/testing/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/client/AbstractKtorHttpClientTest.kt [hunk 4]
```diff
@@ -71,4 +77,24 @@ abstract class AbstractKtorHttpClientTest : AbstractHttpClientTest<HttpRequestBu
...
+    val uri = resolveAddress(path)
+    val responseCode = doRequest(method, uri)
+
+    Assertions.assertThat(responseCode).isEqualTo(200)
+
+    testing.waitAndAssertTraces(
+      Consumer { trace: TraceAssert ->
+        val span = trace.getSpan(0)
+        OpenTelemetryAssertions.assertThat(span).hasKind(SpanKind.CLIENT)
+        Assertions.assertThat(span.endEpochNanos - span.startEpochNanos >= 1000000)
```
Comment (trask) [inline instrumentation/ktor/ktor-2.0/testing/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/client/AbstractKtorHttpClientTest.kt:94]: we typically static import `Assertions` and `OpenTelemetryAssertions` (https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/style-guideline.md#static-imports)

File: instrumentation/ktor/ktor-2.0/testing/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/client/AbstractKtorHttpClientTest.kt (hunk 2 above)
Comment (trask) [reply instrumentation/ktor/ktor-2.0/testing/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/client/AbstractKtorHttpClientTest.kt:82]: @marychatte can you confirm that the change conforms with this part of the spec also? thanks

> If there is any possibility for application code to not fully read the HTTP response (and for the HTTP client library to then have to clean up the HTTP response asynchronously), the HTTP client span SHOULD NOT be ended in this cleanup phase, and instead SHOULD end at some point after the HTTP response headers are fully read (or fail to be read). This avoids the span being ended asynchronously later on at a time which is no longer directly associated with the application code which made the HTTP request.

**Comment by marrek13** [comment]
@marychatte the ticket is here https://youtrack.jetbrains.com/issue/KTOR-7216/Ktor-WebSockets-server-is-not-sending-any-OpenTelemetry-spans-missing-root-span

File: instrumentation/ktor/ktor-2.0/testing/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/client/AbstractKtorHttpClientTest.kt (hunk 2 above)
Comment (marychatte) [reply instrumentation/ktor/ktor-2.0/testing/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/client/AbstractKtorHttpClientTest.kt:82]: @trask, there will be no **later** from `This avoids the span being ended asynchronously later`. Span will be ended right after the body is received. It's not the clean-up phase in the usual sense, we are using `invokeOnCompletion` at the moment of the ending of the body.

Comment (trask) [reply instrumentation/ktor/ktor-2.0/testing/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v2_0/client/AbstractKtorHttpClientTest.kt:81]: thanks, what happens if a user reads the response headers and then decides not to read (ignore) the body?

File: instrumentation/ktor/ktor-2.0/javaagent/build.gradle.kts [hunk 5]
```diff
@@ -9,11 +9,11 @@ muzzle {
   pass {
     group.set("org.jetbrains.kotlinx")
     module.set("ktor-server-core")
-    versions.set("[2.0.0,)")
+    versions.set("[2.3.12,)")
   }
 }
 
-val ktorVersion = "2.0.0"
+val ktorVersion = "2.3.12"
```
Comment (laurit) [inline instrumentation/ktor/ktor-2.0/javaagent/build.gradle.kts:16]: Is this version bump necessary? For javaagent instrumentations we prefer to build/test agains the earliest supported version and test the latest version via the `-PtestLatestDeps=true` flag.

Comment (marychatte) [reply instrumentation/ktor/ktor-2.0/javaagent/build.gradle.kts:16]: No, it's not necessary, I just didn't know about that flag. Thanks, I will revert it

**Review by laurit — APPROVED** [review]

**Pull request merged**